### ISSUE TITLE
Mode ?2027 (RewrapOnResize) conflicts with other terminals

### ISSUE THREAD

**Issue by j4james** [mintty/mintty#1255]
I just noticed that Mintty is using private mode ?2027 to control its `RewrapOnResize` functionality, which I'm assuming is based on Contour's `TextReflow` mode. Unfortunately Contour appears to have switched that mode to ?2028 at some point, and is now using ?2027 for their _Unicode Core_ functionality.

Technically this is Contour's fault, but there are also a number of other terminals now using mode ?2027 for Unicode functionality ([WezTerm](https://github.com/wez/wezterm), [Foot](https://codeberg.org/dnkl/foot), and [Ghostty](https://mitchellh.com/ghostty) are three that I'm aware of), and there's at least one application using that mode (the [vid](https://github.com/erf/vid) editor), so Mintty seems to be in the minority here.

You can argue the case with @christianparpart if you want, but I think it may be best for you to move your `RewrapOnResize` mode to ?2028.

**Comment by mintty** [comment]
I'll switch to 7723 then (23 = ^W).
Where is that new contour mode 2027 described?

**Comment by j4james** [comment]
> I'll switch to 7723 then (23 = ^W). 

Yeah, that's probably best.

> Where is that new contour mode 2027 described?

https://github.com/contour-terminal/terminal-unicode-core

**Comment by mintty** [comment]
I had seen that page already. It does not describe the effect of mode 2027.

**Comment by j4james** [comment]
Quoting from section 5 of the [specification](/contour-terminal/terminal-unicode-core/releases/download/v0.1.0_prerelease_2/terminal-unicode-core-0.1.0-2.pdf):
> 5 Mode Switching
> * CSI ? 2027 h (7) for ensuring conformance to all rules as defined by this specification
> * CSI ? 2027 l (7) for undefined behavior

Basically when mode ?2027 is enabled a terminal should be following the rules laid out in that document. When disabled, it should behave the same way most terminals do now.

**Comment by christianparpart** [comment]
Hey guys. Sorry for the late response. I recently (3 weeks ago) changed jobs, and it takes a little bit of energy :-) .

I am super sorry for the confusion, @mintty. Yes, so 2027 seems to be kind of well established for the Unicode Core Specification. I think we should certainly NOT change the mode for that, since quite some TEs at least are making use of it.

I do not know why I used to use DEC mode 2027 for reflow earlier. Ok, found it. I changed it in 2022-09-13, when incorporating Unicode Core Specification. ;-(
To my defense, I did not think of anyone using what I did there. Text reflow DEC mode was the first custom DEC mode I introduced, AFAIR. I also just realized that we do not document all extensions of Contour in our [fairly new website](https://contour-terminal.org/vt-extensions/).

@mintty is your mode (currently at number) 2027 semantically equal to what we do in Contour, which is: toggle text reflow? I prefer not to chose too high numbers (as you proposed: `7723`). `2028` is (to my knowlege) only used by Contour for reflow.

**Comment by mintty** [comment]
Mintty Reflow disable (now DECRST 7723) is more like a line attribute, it disables reflow on the current line until restored (DECSET 7723). I cannot compare as I do not see an overview of your private escape sequences. Nor did I find a reference to that "Core Spec" on the project page or home page of contour.
How is actually the name "Core Specification" motivated? What's the core here? I do have a few objections against the spec which I'll submit as issues there.
Please do not ever introduce any escape modes beginning with 77. It's the device attributes identification of mintty (ASCII 'M' = 77) following practice of some other terminals. Might be 67 for contour.

**Comment by christianparpart** [comment]
> I cannot compare as I do not see an overview of your private escape sequences

The web site is fairly new, as said, and we'll be filling it over time. I thought Github is enough only until a few months ago. Also motivated by the OSC 60 conflict that xterm generated some time ago.

> Nor did I find a reference to that "Core Spec" on the project page or home page of contour.

We are meant to document all VT extensions here: https://contour-terminal.org/vt-extensions/
But clearly that is incomplete.
Mind, "Vertical Line Markers" is going to change (I should mention it there), because apparently other TEs are having similar features and therfore VT sequences and I simply didn't know about it and somebody told me that the choice (I made long ago) wasn't the best there either.

> How is actually the name "Core Specification" motivated? What's the core here? 

The top of the spec should be pretty delighting on the "core" word, as in, we clearly state, that it's not meant to bring full unicode to the terminal but rather address some core aspects of it that so many other terminal emulators (at the time of writing this spec) has been fighting against adopting, which were proper grapheme cluster support, and all its quirks that came along with it (cell addressing for VT sequences). All I wanted, was to be able to properly render complex emoji on the TE, even the 7-codepoint-long family emoji or anything that has more than one codepoint, including variation selectors (VS15, VS16). It was important to address how the cursor moves forward.

I remember that whenever I wanted to address the Unicode subject to other TE devs, I was always pushed back because it's not meant to be possible (not exact quote here, but this is how the message was received :) ). But the younger generation of power users and developers surely know about graphemes beyond US-ASCII and probably want a little bit of more eye candy (or simply other interesting, but rather complex, code point sequences) in the terminal.

All that was needed is a definition of what to expect as a CLI/TUI app from the TE, and a fairly standard way to query for these semantics. It's not meant to implement full unicode (e.g. explicitly no RTL for a reason), but simply a core, or what I considered to be one. Only later I figured out, that it was mostly about grapheme clusters plus some semantics on cursor placements related to east asian widths.

> Please do not ever introduce any escape modes beginning with 77

I don't intent so :)

> Might be 67 for contour.

I try to keep the numbers as low as possible. I try to not invent any other VT sequence unless absolutely necessary. Specifically DEC mode 2027 (Unicode) as well as [Dark/Light mode detection](https://contour-terminal.org/vt-extensions/color-palette-update-notifications/) are some that I personally find highly valuable, to all TEs and CLI/TUI apps.

p.s.: I'll speed up adding documentation on the website for the missing VT extensions. :)
